### PR TITLE
Fixed for issues#37861

### PR DESCRIPTION
**Pull request by v-hongli1**
relate to [issue#37861](https://github.com/Azure/azure-sdk-for-java/issues/37861)

# Description

1. Add `userData` support in create and update APIs.
2. Add test case `VirtualMachineOperationsTests#canCreateAndUpdateVirtualMachineWithUserData`.

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/resourcemanager/azure-resourcemanager-compute/src/main/java/com/azure/resourcemanager/compute/models/VirtualMachine.java [hunk 1]
```diff
@@ -441,6 +441,13 @@ Mono<RunCommandResult> runShellScriptAsync(
      */
     DeleteOptions networkInterfaceDeleteOptions(String networkInterfaceId);
 
+    /**
+     * Gets the user data for the virtual machine.
+     *
+     * @return the user data for the virtual machine.
+     */
+    String userData();
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-compute/src/main/java/com/azure/resourcemanager/compute/models/VirtualMachine.java:449]: @XiaofeiCao should we use `byte[]` for both setter and getter?

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java [hunk 2]
```diff
@@ -104,6 +108,38 @@ protected void cleanUpResources() {
         }
     }
 
+    @Test
+    @Disabled("The `userData` is not returned, so can not use `Assertions.assertEquals` be determine whether the returned `userData` is correctly.")
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:112]: Enabled the test anyway.

If userData is not returned, we should not provide the getter.

Some data is secret and not returned.

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:112]: Agree, getter is not needed.

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java [hunk 3]
```diff
@@ -104,6 +108,38 @@ protected void cleanUpResources() {
...
+            .define(vmName)
+            .withRegion(region)
+            .withNewResourceGroup(rgName)
+            .withNewPrimaryNetwork("10.0.0.0/28")
+            .withPrimaryPrivateIPAddressDynamic()
+            .withoutPrimaryPublicIPAddress()
+            .withPopularWindowsImage(KnownWindowsVirtualMachineImage.WINDOWS_SERVER_2012_R2_DATACENTER)
+            .withAdminUsername("Foo12")
+            .withAdminPassword(password())
+            .withUnmanagedDisks()
```
[the remark this answers is not in the document]
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java]: 1. The getter of `userData`  is not needed, so remove it.
2. Enabled the test case `VirtualMachineOperationsTests#canCreateAndUpdateVirtualMachineWithUserData`.
3. Used managed disk instead.

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java [hunk 4]
```diff
@@ -104,6 +108,34 @@ protected void cleanUpResources() {
         }
     }
 
+    @Test
+    public void canCreateAndUpdateVirtualMachineWithUserData() {
+        String userDataForCreate = new String(Base64.getEncoder().encode(
+            UUID.randomUUID().toString().toUpperCase(Locale.ROOT).getBytes(StandardCharsets.UTF_8)));
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:114]: randomUUID cannot be created this way. The random data need to be recorded in the test record.

Use `ResourceManagerUtil.randomUuid`.

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:114]: Or just use some non-random string. If the content is not important.

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java [hunk 5]
```diff
@@ -104,6 +108,34 @@ protected void cleanUpResources() {
...
+        // Create
+        VirtualMachine vm = computeManager
+            .virtualMachines()
+            .define(vmName)
+            .withRegion(region)
+            .withNewResourceGroup(rgName)
+            .withNewPrimaryNetwork("10.0.0.0/28")
+            .withPrimaryPrivateIPAddressDynamic()
+            .withoutPrimaryPublicIPAddress()
+            .withPopularWindowsImage(KnownWindowsVirtualMachineImage.WINDOWS_SERVER_2012_R2_DATACENTER)
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:127]: Can we use a later windows version? 2012 seems too old.

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java (hunk 2 above)
Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java]: please check https://learn.microsoft.com/en-us/azure/virtual-machines/user-data

> User data can be queried via GET VM/VMSS API with $expand option.

seems userData can be get, but need some expand option?

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java (hunk 5 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java]: 1. The value of  `userData` changed to two different non-random strings.
2. The version of windows virtual machine image changed to `2019-datacenter-gensecond`.

**Comment by weidongxu-microsoft** [comment]
see https://github.com/Azure/azure-sdk-for-java/pull/37923#discussion_r1416660945

**Review by weidongxu-microsoft — APPROVED** [review]
Please also wait for Xiaofei's review.

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java [hunk 6]
```diff
@@ -104,6 +109,39 @@ protected void cleanUpResources() {
...
+            .withNewDataDisk(127)
+            .withSize(VirtualMachineSizeTypes.fromString("Standard_D2a_v4"))
+            .withUserData(userDataForCreate)
+            .create();
+        Response<VirtualMachineInner> response = computeManager.serviceClient().getVirtualMachines()
+            .getByResourceGroupWithResponse(rgName, vmName, InstanceViewTypes.USER_DATA, Context.NONE);
+        Assertions.assertEquals(userDataForCreate, response.getValue().userData());
+
+        // Update
+        computeManager.serviceClient().getVirtualMachines()
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:138]: Why use serviceClient method for updating VirtualMachine? I guess there can be:
```java
vm.update()
    .withUserData(userDataForUpdate)
    .apply()
```
right?

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:138]: Using `vm.update().withUserData(userDataForUpdate).apply()` cannot update the `VirtualMachine` service in real time, so use the serviceClient method to updating the `VirtualMachine`.

**Review by weidongxu-microsoft — CHANGES_REQUESTED** [review]
https://github.com/Azure/azure-sdk-for-java/pull/37923#discussion_r1418353659
this need to be fixed. update().apply() had to work.

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java (hunk 6 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:138]: We did some parameter copying from `innerModel` to update parameter.
https://github.com/Azure/azure-sdk-for-java/blob/04f9b83c351f6d817da6c73308dc3a37c64d03db/sdk/resourcemanager/azure-resourcemanager-compute/src/main/java/com/azure/resourcemanager/compute/implementation/VirtualMachineImpl.java#L2764-L2781

 Probably need to add additional `userData` copy here.

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java]: It has been fixed in the new version and `userData` can be updated.

**Review by XiaofeiCao — APPROVED** [review]

**Pull request merged**